sdk/python/lib/pulumi/runtime/resource.py

```diff
@@ -138,8 +139,62 @@ async def prepare_resource(res: 'Resource',
     )
 
 
-def read_resource(res: 'CustomResource', ty: str, name: str, props: 'Inputs', opts: 'ResourceOptions') -> None:
+def resource_output(res: 'Resource') -> Tuple[Callable[[Any, bool, bool, Optional[Exception]], None], 'Output']:
     from .. import Output  # pylint: disable=import-outside-toplevel
+
+    value_future: asyncio.Future[Any] = asyncio.Future()
+    known_future: asyncio.Future[bool] = asyncio.Future()
+    secret_future: asyncio.Future[bool] = asyncio.Future()
+
+    def resolve(value: Any, known: bool, secret: bool, exn: Optional[Exception]):
+        if exn is not None:
+            value_future.set_exception(exn)
+            known_future.set_exception(exn)
+            secret_future.set_exception(exn)
+        else:
+            value_future.set_result(value)
+            known_future.set_result(known)
+            secret_future.set_result(secret)
+
+    return (resolve, Output({res}, value_future, known_future, secret_future))
+
+
+def resolved_resource_output(res: 'Resource', value: Any, known: bool, secret: bool) -> 'Output':
+    (resolve, output) = resource_output(res)
+    resolve(value, known, secret, None)
+    return output
+
+
+def get_resource(res: 'Resource', custom: bool, urn: str) -> None:
+    log.debug(f"getting resource: urn={urn}")
+
+    result = invoke("pulumi:pulumi:getResource", {"urn": urn})
+
+    # Same as below, we initialize the URN property on the resource, which will always be resolved.
+    log.debug("preparing get resource for RPC")
+
+    res.__dict__["urn"] = resolved_resource_output(res, urn, True, False)
+
+    # If this is a custom resource, resolve the ID.
+    if custom:
+        id = result.value["id"]
+        id_known = bool(id)
+
+        res.__dict__["id"] = resolved_resource_output(res, id, id_known, False)
+
+    # Create resolved outputs for each property on res.
+    for key, value in result.value["state"].items():
```

Comment:
Do we also need to handle the case where there may be optional outputs that aren't present in the state, but do exist as properties on the resource?

https://github.com/pulumi/pulumi/blob/f22b47e54f39312b9b13ea359d3d523ee42e7a82/sdk/python/lib/pulumi/runtime/rpc.py#L633-L638

Reply:
Hmmm, we do. That's a very good point.